scriptmodules/supplementary/emulationstation/configscripts/mupen64plus.sh

```diff
@@ -0,0 +1,266 @@
+                if echo "$ini_value" | grep -q "("; then
+                    value="${ini_value}${input_id})"
+                # if there is already a "+" value add "-" value
+                elif echo "$ini_value" | grep -q ")"; then
+                    value="button(${input_id},${ini_value}"
+                elif [[ "$dir" == "Up" ]]; then
+                    value="button(${input_id},"
+                elif [[ "$dir" == "Right" ]]; then
+                    value="${input_id})"
+                elif [[ "$dir" == "Down" ]]; then
+                    value="${input_id})"
+                elif [[ "$dir" == "Left" ]]; then
+                    value="button(${input_id},"
+                fi
+            fi
+            value="button(${input_id}) ${ini_value}"
+        fi
+
+        iniSet "$key" "$value"
+    done
+}
+
+function onend_mupen64plus_joystick() {
+    local device_type=$1
+    local device_name=$2
+    
+    # analog stick sanity check
+    # replace Axis value with DPAD values if there is no Axis 
+    # device setup
+    local bind;
+    if ! grep -q -e "X Axis" -e "Y Axis" /tmp/mp64tempconfig.cfg; then
+        iniGet "DPad U"
+        bind=`echo "$ini_value" | sed -e "s/)/,/g"`
+        iniGet "DPad D"
+        ini_value=`echo "$ini_value" | sed -e "s/button(//g"`
+        ini_value=`echo "$ini_value" | sed -e "s/hat(//g"`
+        ini_value=`echo "$ini_value" | sed -e "s/axis(//g"`
+        bind="${bind}${ini_value}"
+        iniSet "Y Axis" "$bind"
+        iniSet "DPad U" ""
+        iniSet "DPad D" ""
+        
+        iniGet "DPad L"
+        bind=`echo "$ini_value" | sed -e "s/)/,/g"`
+        iniGet "DPad R"
+        ini_value=`echo "$ini_value" | sed -e "s/button(//g"`
+        ini_value=`echo "$ini_value" | sed -e "s/hat(//g"`
+        ini_value=`echo "$ini_value" | sed -e "s/axis(//g"`
+        bind="${bind}${ini_value}"
+        iniSet "X Axis" "$bind"
+        iniSet "DPad L" ""
+        iniSet "DPad R" ""
+    fi
+    
+    # remove analog values if "Axis" is no analog device
+    iniGet "X Axis"
+    if echo "$ini_value" | grep -q -e "button" -e "hat"; then
+        iniDel "AnalogDeadzone"
+        iniDel "AnalogPeak"
+    fi
+    iniGet "Y Axis"
+    if echo "$ini_value" | grep -q -e "button" -e "hat"; then
+        iniDel "AnalogDeadzone"
+        iniDel "AnalogPeak"
+    fi
+    
+    echo "; ${device_name}_END " >> /tmp/mp64tempconfig.cfg
+    echo "" >> /tmp/mp64tempconfig.cfg
+
+    # abort if old device config cannot be deleted.
+    # keep original mupen64plus-input-sdl configs.
+    local file="/opt/retropie/configs/n64/InputAutoCfg.ini"
+    if [[ -f "$file" ]]; then
+        # backup current config file
+        cp "$file" "${file}.bak"
+        sed -i /"${device_name}_START"/,/"${device_name}_END"/d "$file"
+        if grep -q "$device_name" "$file"; then
+            return
+        fi
+    else
+        cat > "$file" << _EOF_
+; InputAutoCfg.ini for Mupen64Plus SDL Input plugin
+
+[Keyboard]
```

Comment:
keyboard input is hardcoded or ? if keyboard is configurable we should try and configure that in the case that the user uses a keyboard to configure ES (although this could come later).